File: data/derived_cot/rq2_eval/translation/r1/_converted_code/java_to_py/NLPDataProcessor2.py
import re

class NLPDataProcessor2:

    class WordFrequency:
        __slots__ = ('word', 'frequency')

        def __init__(self, word, frequency):
            self.word = word
            self.frequency = frequency

        def __eq__(self, other):
            if self is other:
                return True
            if not isinstance(other, NLPDataProcessor2.WordFrequency):
                return False
            return self.frequency == other.frequency and self.word == other.word

        def __hash__(self):
            return hash((self.word, self.frequency))

        def __str__(self):
            return f"WordFrequency{{word='{self.word}', frequency={self.frequency}}}"

        @staticmethod
        def byFrequencyThenWord():

            def comparator(wf1, wf2):
                if wf1.frequency != wf2.frequency:
                    return wf2.frequency - wf1.frequency
                return (wf1.word > wf2.word) - (wf1.word < wf2.word)
            return comparator

    def process_data(self, stringList):
        pattern = re.compile('[^a-zA-Z\\s]')
        words_list = []
        for s in stringList:
            processed = pattern.sub('', s.lower())
            if processed == '':
                words = []
            else:
                words = re.split('\\s+', processed)
                while words and words[-1] == '':
                    words.pop()
            words_list.append(words)
        return words_list

    def calculate_word_frequency(self, wordsList):
        frequency_map = {}
        order_map = {}
        index = 0
        for words in wordsList:
            for word in words:
                if word not in frequency_map:
                    frequency_map[word] = 0
                    order_map[word] = index
                    index += 1
                frequency_map[word] += 1
        word_frequencies = []
        for word, freq in frequency_map.items():
            if freq > 1 or word == '%%%':
                word_frequencies.append(self.WordFrequency(word, freq))
        word_frequencies.sort(key=lambda wf: (-wf.frequency, order_map[wf.word]))
        return word_frequencies

    def process(self, stringList):
        wordsList = self.process_data(stringList)
        return self.calculate_word_frequency(wordsList)

File: data/derived_cot/rq1_traces/translation/glm5.1/java_to_cpp/code_output/NLPDataProcessor2.cpp
#include <vector>
#include <string>
#include <unordered_map>
#include <algorithm>
#include <regex>
#include <cctype>
#include <functional>

class NLPDataProcessor2 {
public:
    class WordFrequency {
    private:
        std::string word_;
        int frequency_;

    public:
        WordFrequency(const std::string& word, int frequency)
            : word_(word), frequency_(frequency) {}

        std::string getWord() const { return word_; }
        int getFrequency() const { return frequency_; }

        std::string toString() const {
            return "WordFrequency{word='" + word_ + "', frequency=" + std::to_string(frequency_) + '}';
        }

        bool equals(const WordFrequency& that) const {
            return frequency_ == that.frequency_ && word_ == that.word_;
        }

        bool operator==(const WordFrequency& that) const {
            return equals(that);
        }

        bool operator!=(const WordFrequency& that) const {
            return !equals(that);
        }

        size_t hashCode() const {
            size_t h1 = std::hash<std::string>()(word_);
            size_t h2 = std::hash<int>()(frequency_);
            return h1 ^ (h2 << 1);
        }

        static std::function<bool(const WordFrequency&, const WordFrequency&)> byFrequencyThenWord() {
            return [](const WordFrequency& wf1, const WordFrequency& wf2) {
                if (wf1.getFrequency() != wf2.getFrequency()) {
                    return wf1.getFrequency() > wf2.getFrequency();
                }
                return wf1.getWord() < wf2.getWord();
            };
        }
    };

private:
    // Replicates Java's String.split(regex) which discards trailing empty strings
    static std::vector<std::string> javaSplit(const std::string& str, const std::regex& pattern) {
        std::vector<std::string> result;
        std::sregex_token_iterator it(str.begin(), str.end(), pattern, -1);
        std::sregex_token_iterator end;
        for (; it != end; ++it) {
            result.push_back(it->str());
        }
        while (!result.empty() && result.back().empty()) {
            result.pop_back();
        }
        return result;
    }

public:
    std::vector<std::vector<std::string>> processData(const std::vector<std::string>& stringList) {
        std::vector<std::vector<std::string>> wordsList;
        std::regex pattern("[^a-zA-Z\\s]");

        for (const std::string& str : stringList) {
            std::string lowerStr = str;
            std::transform(lowerStr.begin(), lowerStr.end(), lowerStr.begin(),
                           [](unsigned char c) { return std::tolower(c); });
            std::string processedString = std::regex_replace(lowerStr, pattern, "");

            std::vector<std::string> words;
            if (!processedString.empty()) {
                std::regex wsRegex("\\s+");
                words = javaSplit(processedString, wsRegex);
            }
            wordsList.push_back(words);
        }
        return wordsList;
    }

    std::vector<WordFrequency> calculateWordFrequency(const std::vector<std::vector<std::string>>& wordsList) {
        std::unordered_map<std::string, int> frequencyMap;
        std::unordered_map<std::string, int> orderMap;
        std::vector<std::string> insertionOrder;
        int index = 0;

        for (const auto& words : wordsList) {
            for (const std::string& word : words) {
                if (frequencyMap.find(word) == frequencyMap.end()) {
                    orderMap[word] = index++;
                    insertionOrder.push_back(word);
                }
                frequencyMap[word]++;
            }
        }

        std::vector<WordFrequency> wordFrequencies;
        for (const auto& word : insertionOrder) {
            int frequency = frequencyMap[word];
            if (frequency > 1 || word == "%%%") {
                wordFrequencies.push_back(WordFrequency(word, frequency));
            }
        }

        std::sort(wordFrequencies.begin(), wordFrequencies.end(),
            [&orderMap](const WordFrequency& wf1, const WordFrequency& wf2) {
                if (wf1.getFrequency() != wf2.getFrequency()) {
                    return wf1.getFrequency() > wf2.getFrequency();
                }
                return orderMap[wf1.getWord()] < orderMap[wf2.getWord()];
            });

        return wordFrequencies;
    }

    std::vector<WordFrequency> process(const std::vector<std::string>& stringList) {
        std::vector<std::vector<std::string>> wordsList = processData(stringList);
        return calculateWordFrequency(wordsList);
    }
};